File: plugin/src/main/java/com/jtprince/bingo/plugin/automarking/AutoMarkTrigger.java
package com.jtprince.bingo.plugin.automarking;

import com.jtprince.bingo.plugin.Space;
import com.jtprince.bingo.plugin.automarking.itemtrigger.ItemTrigger;

import java.util.Collection;
import java.util.HashSet;

public abstract class AutoMarkTrigger {
    private final Space space;

    protected AutoMarkTrigger(Space space) {
        this.space = space;
    }

    protected AutoMarkTrigger() {
        this.space = null;
    }

    public static Collection<AutoMarkTrigger> createAllTriggers(Space space) {
        HashSet<AutoMarkTrigger> set = new HashSet<>();

        set.addAll(ItemTrigger.createTriggers(space.goalId));
        set.addAll(EventTrigger.createTriggers(space));
        set.addAll(OccasionalTrigger.createTriggers(space));

        return set;
    }

    public Space getSpace() {
        return this.space;
    }

    public abstract void destroy();
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/automark/AutoMarkTrigger.kt
package com.jtprince.bingo.kplugin.automark

import com.jtprince.bingo.kplugin.board.SetVariables
import com.jtprince.bingo.kplugin.board.Space
import com.jtprince.bingo.kplugin.game.PlayerManager
import com.jtprince.bingo.kplugin.player.BingoPlayer

typealias AutoMarkCallback = (BingoPlayer, spaceId: Int, fulfilled: Boolean) -> Unit

abstract class AutoMarkTrigger(
    val goalId: String,
    val spaceId: Int,
    val variables: SetVariables,
    val callback: AutoMarkCallback,
    val playerManager: PlayerManager,
) {
    companion object Creator {
        fun createForGoal(goalId: String, spaceId: Int, variables: SetVariables,
                          playerManager: PlayerManager, callback: AutoMarkCallback
        ): Collection<AutoMarkTrigger> {
            return (EventTrigger.createEventTriggers(goalId, spaceId, variables, playerManager, callback)
                + ItemTrigger.createItemTriggers(goalId, spaceId, variables, playerManager, callback))
        }
    }

    /**
     * If (and only if) revertible is true, the AutoMarkCallback may be called with
     * fulfilled == false. In this case, the callback should put the goal into a "reverted" state.
     */
    abstract val revertible: Boolean

    abstract fun destroy()
}
